compiler/rustc_transmute/src/layout/tree.rs

```diff
@@ -0,0 +1,479 @@
+                ty::Int(I128) | ty::Uint(U128) => Ok(Self::number(16)),
+                ty::Int(Isize) | ty::Uint(Usize) => {
+                    Ok(Self::number(target.pointer_size.bytes_usize()))
+                }
+
+                ty::Tuple(members) => {
+                    if members.len() == 0 {
+                        Ok(Tree::unit())
+                    } else {
+                        Err(Err::Unspecified)
+                    }
+                }
+
+                ty::Array(ty, len) => {
+                    let len = len.try_eval_usize(tcx, ParamEnv::reveal_all()).unwrap();
+                    let elt = Tree::from_ty(*ty, tcx)?;
+                    Ok(std::iter::repeat(elt)
+                        .take(len as usize)
+                        .fold(Tree::unit(), |tree, elt| tree.then(elt)))
+                }
+
+                ty::Adt(adt_def, substs_ref) => {
+                    use rustc_middle::ty::AdtKind;
+
+                    // If the layout is ill-specified, halt.
+                    if !(adt_def.repr().c() || adt_def.repr().int.is_some()) {
+                        return Err(Err::Unspecified);
+                    }
+
+                    // Compute a summary of the type's layout.
+                    let layout_summary = LayoutSummary::from_ty(ty, tcx)?;
+
+                    // The layout begins with this adt's visibility.
+                    let vis = Self::def(Def::Adt(*adt_def));
+
+                    // And is followed the layout(s) of its variants
+                    Ok(vis.then(match adt_def.adt_kind() {
+                        AdtKind::Struct => Self::from_repr_c_variant(
+                            ty,
+                            *adt_def,
+                            substs_ref,
+                            &layout_summary,
+                            None,
+                            adt_def.non_enum_variant(),
+                            tcx,
+                        )?,
+                        AdtKind::Enum => {
+                            tracing::trace!(
+                                adt_def = ?adt_def,
+                                "treeifying enum"
+                            );
+                            let mut tree = Tree::uninhabited();
+
+                            for (idx, discr) in adt_def.discriminants(tcx) {
+                                tree = tree.or(Self::from_repr_c_variant(
+                                    ty,
+                                    *adt_def,
+                                    substs_ref,
+                                    &layout_summary,
+                                    Some(discr),
+                                    adt_def.variant(idx),
+                                    tcx,
+                                )?);
+                            }
+
+                            tree
+                        }
+                        AdtKind::Union => {
```

Comment:
Can we actually say anything about unions? I was under the impression the UCG working group said unions are just bags of bytes with no rules

Reply:
Hey @RalfJung, my understanding is you've been active in discussions around unions. The essence of this PR is that it introduces the compiler-implemented trait `TransmutableFrom`[^1]. The bound `Dst: TransmutableFrom<Src>` holds for given `Src` and `Dst` if the layouts of `Src` and `Dst` are specified, and all possible values of `Src` are also bit-valid instances of `Dst`. The implication of satisfying `Dst: TransmutableFrom<Src>` is that you may soundly (without any runtime checks or other preconditions) transmute any arbitrary value of `Src` into `Dst`.

[^1]: I'm simplifying slightly. The trait is actually currently called `BikeshedIntrinsicFrom` and takes in a few other parameters not relevant to this discussion.

I've been operating under the assumption it is sound for a library to provide this:
```rust
#[repr(C)]
pub union Union {
    // n.b.: this field is public
    pub f: bool,
}

impl Union {
    // n.b.: this method is safe
    pub fn into_bool(self) -> bool {
        // i believe this method is sound, because
        // all bit-valid instances of `Union` are
        // bit-valid instances of `bool`
        assert_is_transmutable::<Union, bool>();
        unsafe { self.f }
    }
}

fn assert_is_transmutable<Src, Dst>()
where
    Dst: mem::TransmutableFrom<Src>
{}
```
...because I've been assuming it's **NOT** sound to ever provide this:
```rust
pub fn crime(mut u: Union) -> Union {
    unsafe {
        let f = &mut u.f as *mut _ as *mut u8;
        *f = 3;
    }
    u
}
```
Are my assumptions correct?

If it `fn crime()` *is* sound (e.g., because unions are specified to be only bags of possibly-initialized, arbitrary bytes), then I should revise my implementation such that `assert_is_transmutable::<Union, bool>();` is a compile-error.

Reply:
The library defining `Union` can provide *either* `into_bool` or `crime`, but not both. Either choice is itself sound, but providing both of them in combination is (obviously) not sound.

This is similar to, e.g., a library defining a `struct Even(i32)` can soundly provide
- EITHER a constructor that turns an arbitrary `i32` into an `Even`
- OR a function that calls `unreachable_unchecked` when the field value is odd

but not both.

So, it is a choice of the library author whether they want the invariant for their union to be "must contain a `bool`" (and offer `into_bool`) or not. I don't think the safe transmute infrastructure should *implicitly* make such a choice for the library author. (But it might be useful to let library authors explicitly opt-in to such a choice. The details of what exactly that choice would mean are subtle though and should be discussed carefully. I'm happy to provide advice here.)

Reply:
Thanks for taking the time to clarify! Given this, I believe the correct course of action for my PR is to treat all unions as if they implicitly have a variant of type `()`. Consequently:
- virtually any `Src` is transmutable into any `Union` (because anything is a bit-valid instance of uninit bytes)
- a `Union` is *only* safely transmutable into `Dst`s consisting only of uninit bytes (because nothing except uninit bytes are valid instances of uninit bytes)

Are the validity rules of unions a settled matter? I understand a few alternatives have been discussed, but I don't have a sense of what the tradeoffs are of each. From where I stand, it'd be *very* convenient to have a stricter validity model, as that'd let me do much more with statically analyzing operations involving unions.

Under the more restrictive rule that the n<sup>th</sup> byte of a union satisfies the bit validity requirements of the n<sup>th</sup> byte of any of its variants, it's possible to construct an enum and unions with equal layouts (and to confirm this to be true with `TransmuteFrom`); e.g.:
```rust
#[derive(Copy, Clone)] #[repr(u8)] enum Ox00 { Value = 0x00 }
#[derive(Copy, Clone)] #[repr(u8)] enum Ox01 { Value = 0x01 }
#[derive(Copy, Clone)] #[repr(u8)] enum Ox02 { Value = 0x02 }

#[repr(u8)]
enum Enum {
    False = 0x00,
    True = 0x01,
}

#[repr(C)]
union Union {
    False: Ox00,
    True: Ox01,
}

// statically assert that these items have identical layouts:
is_transmutable::<Enum, Union>();
is_transmutable::<Union, Enum>();
```
And, under this stricter rule, I can define a macro (or even a language feature) implementing safe union access ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4dcdd8065bd24b1cbffc2cb15abcc5f6)):
```rust
/// Safely accesses a union field, if possible; otherwise
/// emits a compile error.
macro access($union:expr => $field:ident) {
    if true {
        unsafe {
            $union.$field
        }
    } else {
        fn is_transmutable<Src, Dst>(_: Src) -> Dst
        where
            Dst: TransmutableFrom<Src>
        {
            loop {}
        }

        is_transmutable($union)
    }
}

#[repr(C)]
union Union {
    i32: i32,
    f32: f32,    
}

fn main() {
    let u = Union { i32: 42i32 };
    let f = access!(u => f32);
    println!("{:?}", f);
}
```
And, under this stricter rule, library authors *already* have a language-level mechanism for indicating that their enum *should* admit arbitrary bytes, if they so wish: just add a `()` variant.

By contrast, under the rule that unions *implicitly* admit arbitrary bytes, none of this is possible. Although library authors can still certainly make a choice on this matter (by keeping the variants of their unions private, and by committing themselves to stick to the convention of providing either `into_bool` or `crime` but not both), their choice will generally be invisible to static analysis.

I might also add that it feels very odd (perhaps misleading?) to apply the keyword "`union`" to something whose layout *isn't* simply the union of its variants.

Anyways, if there's anything you can link me to summarizing the tradeoffs of the various validity proposals, I'd be grateful. I'm sure there are downsides of the stricter validity rule that I'm not aware of.

Reply:
> Are the validity rules of unions a settled matter?

No. So I'd appreciate if safe-transmute didn't settle them by making irreversible decisions here. ;)

You raise some good points that have been discussed, for example in
- https://github.com/rust-lang/unsafe-code-guidelines/issues/352
- https://github.com/RalfJung/minirust/pull/14
- [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/sum.2Fbag-o-bytes.20union.20repr)

Reply:
> Given this, I believe the correct course of action for my PR is to treat all unions as if they implicitly have a variant of type (). Consequently:

I cannot say whether any of this is correct until seeing your reply to [this comment](https://github.com/rust-lang/rust/pull/92268#issuecomment-1190714566). Using the WG-UCG's interpretation of soundness, "virtually any Src is [soundly] transmutable into any Union" is wrong, though maybe you meant another kind of transmutability. Also, *any* type can be soundly transmuted to a type that is known to accept arbitrary data for all its bytes, that follows trivially from "it is sound to convert Src into Dst (via transmutation) if an only if the safety invariant of Src implies that of Dst".

Reply:
> No. So I'd appreciate if safe-transmute didn't settle them by making irreversible decisions here. ;)

I won't. :-) This is only an MCP — not an RFC — and, actually,  I'm hopeful it might be a useful tool for illustrating these sorts of layout decisions. I'll review the discussions you linked, and perhaps implement a few of the proposals there.